domain/secret/state/state.go

```diff
@@ -451,6 +448,53 @@ VALUES ($secretID.id)`
 	return nil
 }
 
+// GetSecretOwner returns the owner of the secret with the given URI, returning an error satisfying
+// [secreterrors.SecretNotFound] if the secret does not exist.
+func (st State) GetSecretOwner(ctx domain.AtomicContext, uri *coresecrets.URI) (coresecrets.Owner, error) {
+	input := secretID{ID: uri.ID}
+	stmt, err := st.Prepare(`
+SELECT
+       (so.owner_kind,
+        so.owner_id) AS (&secretOwner.*)
+FROM   secret_metadata sm
+       LEFT JOIN (
+          SELECT $ownerKind.model_owner_kind AS owner_kind, '' AS owner_id,  secret_id
+          FROM   secret_model_owner so
+          UNION
+          SELECT $ownerKind.application_owner_kind AS owner_kind, application.uuid AS owner_id,  secret_id
+          FROM   secret_application_owner so
+          JOIN   application
+          WHERE  application.uuid = so.application_uuid
+          UNION
+          SELECT $ownerKind.unit_owner_kind AS owner_kind, unit_uuid AS owner_id,  secret_id
+          FROM   secret_unit_owner so
+          JOIN   unit
+          WHERE  unit.uuid = so.unit_uuid
+       ) so ON so.secret_id = sm.secret_id
+WHERE  sm.secret_id = $secretID.id
+`, secretID{}, secretOwner{}, ownerKindParam)
+	if err != nil {
+		return coresecrets.Owner{}, errors.Trace(err)
+	}
+
+	var result []secretOwner
+	err = domain.Run(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, input, ownerKindParam).GetAll(&result)
+		if errors.Is(err, sqlair.ErrNoRows) || len(result) == 0 {
+			return fmt.Errorf("secret %q not found%w", uri, errors.Hide(secreterrors.SecretNotFound))
+		}
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return coresecrets.Owner{}, errors.Trace(err)
+	}
+	owner := result[0]
+	return coresecrets.Owner{
+		ID:   owner.OwnerID,
```

Comment:
This is overloading the ID of secrets.Owner. The ID is for the natural key of the owner entity. Here we're setting it to a UUID and in the services code there's some code which uses Owner{} with ID set to the app or unit name and then other places where the ID is set to app or unit UUID. If we want to have a struct with UUID for use in this domain, we should declare one in this domain instead of using the one from core.